File: libs/go/pkg/autonomy/compliance.go
package autonomy

import (
	"errors"
	"regexp"
	"strings"
)

// ComplianceEngine is the Phase 08 policy-check contract.
//
//	true, nil   = passed the development rules
//	false, nil  = policy violation detected
//	false, err  = check unavailable / malformed
//
// UNKNOWN/ERROR ≠ COMPLIANT.
type ComplianceEngine interface {
	Check(text string) (bool, error)
}

// DevelopmentCompliance is a deterministic DEVELOPMENT / TEST policy engine.
// It is not legal compliance certification, complete privacy detection,
// comprehensive moderation, or jurisdiction-aware legal analysis.
//
// Supported detections: prohibited phrase list; simple email; US-SSN-like pattern.
// Known false positives: any string containing those patterns.
// Known false negatives / unsupported: phone, address, passport, national IDs, images, multilingual PII.
type DevelopmentCompliance struct{}

const ComplianceSafeFixture = "DEV_TRUTH_FIXTURE: local observation"

var (
	ErrComplianceUnavailable = errors.New("COMPLIANCE_UNAVAILABLE")
	complianceProhibited     = []string{"prohibited:unlicensed-medical-claim", "bypass compliance"}
	emailRE                  = regexp.MustCompile(`(?i)[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}`)
	ssnRE                    = regexp.MustCompile(`\b\d{3}-\d{2}-\d{4}\b`)
)

func (DevelopmentCompliance) Kind() string { return "DEVELOPMENT_POLICY_ENGINE" }

func (DevelopmentCompliance) Check(text string) (bool, error) {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return false, ErrComplianceUnavailable
	}
	lower := strings.ToLower(trimmed)
	for _, p := range complianceProhibited {
		if strings.Contains(lower, p) {
			return false, nil
		}
	}
	if emailRE.MatchString(trimmed) || ssnRE.MatchString(trimmed) {
		return false, nil
	}
	return true, nil
}

File: apps/web/lib/autonomy-control/compliance.ts
/**
 * DEVELOPMENT / TEST compliance rule engine.
 * Not legal certification, not comprehensive PII, not jurisdiction-aware.
 *
 * Supported detections: explicit prohibited phrase; simple email; US-SSN-like \d{3}-\d{2}-\d{4}.
 * Known false positives: any text containing those patterns.
 * Known false negatives / unsupported: phone, address, passport, national IDs, images, multilingual PII.
 */

export const COMPLIANCE_SAFE_FIXTURE = "DEV_TRUTH_FIXTURE: local observation";

const PROHIBITED = ["prohibited:unlicensed-medical-claim", "bypass compliance"];
const EMAIL = /[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i;
const SSN = /\b\d{3}-\d{2}-\d{4}\b/;

export class ComplianceUnavailable extends Error {
  readonly code = "COMPLIANCE_UNAVAILABLE";
  constructor(message: string) {
    super(message);
    this.name = "ComplianceUnavailable";
  }
}

export type ComplianceResult = { ok: boolean; error?: ComplianceUnavailable };

export class DevComplianceEngine {
  readonly kind = "DEVELOPMENT_POLICY_ENGINE";

  check(text: string): ComplianceResult {
    const trimmed = (text ?? "").trim();
    if (!trimmed) return { ok: false, error: new ComplianceUnavailable("empty content cannot be checked") };
    const lower = trimmed.toLowerCase();
    if (PROHIBITED.some((p) => lower.includes(p))) return { ok: false };
    if (EMAIL.test(trimmed) || SSN.test(trimmed)) return { ok: false };
    return { ok: true };
  }
}
